File: 13.gRPC-and-Inter-Service-Communication-for-Microservices/code/typescript/authentication_security_9.ts
import * as grpc from '@grpc/grpc-js';
import * as fs from 'fs';

// ===== SERVER over TLS =====
// const serverCerts = fs.readFileSync('server.crt');
// const serverKey = fs.readFileSync('server.key');
// const creds = grpc.ServerCredentials.createSsl(null, [{
//     cert_chain: serverCerts,
//     private_key: serverKey
// }]);

// every connection is now encrypted
// const server = new grpc.Server();
// server.bindAsync('0.0.0.0:443', creds, ...);

// ===== CLIENT over TLS =====
// const rootCerts = fs.readFileSync('ca.crt');
// const tlsCreds = grpc.credentials.createSsl(rootCerts); // trust this CA

const tlsCreds = grpc.credentials.createInsecure(); // Dummy for compilation
const client = new UserServiceClient('api.example.com:443', tlsCreds);

// ===== Per-call token (combine with TLS) =====
// Implement CallCredentials so a fresh token rides on every call:
const token = "my-secret-token";
const tokenCreds = grpc.credentials.createFromMetadataGenerator((args, cb) => {
    const md = new grpc.Metadata();
    md.add('authorization', 'Bearer ' + token);
    cb(null, md);
});

// combine transport and per-call credentials
const combinedCreds = grpc.credentials.combineChannelCredentials(
    tlsCreds,
    tokenCreds
);

const clientWithToken = new UserServiceClient('api.example.com:443', combinedCreds);

class UserServiceClient { constructor(a:any, b:any){} }

File: 13.gRPC-and-Inter-Service-Communication-for-Microservices/code/rust/authentication_security_9.rs
use tonic::transport::{Certificate, Identity, Server, ClientTlsConfig, Channel};
use tonic::metadata::MetadataValue;
use std::fs;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // ===== SERVER over TLS =====
    /* 
    let cert = fs::read("server.crt")?;
    let key = fs::read("server.key")?;
    let identity = Identity::from_pem(cert, key);
    
    let tls_config = tonic::transport::ServerTlsConfig::new()
        .identity(identity);

    // every connection is now encrypted
    Server::builder()
        .tls_config(tls_config)?
        .add_service(svc)
        .serve(addr)
        .await?;
    */

    // ===== CLIENT over TLS =====
    /*
    let root_ca = fs::read("ca.crt")?;
    let cert = Certificate::from_pem(root_ca); // trust this CA
    
    let tls_config = ClientTlsConfig::new()
        .ca_certificate(cert)
        .domain_name("api.example.com");

    let channel = Channel::from_static("https://api.example.com:443")
        .tls_config(tls_config)?
        .connect()
        .await?;

    let mut client = UserServiceClient::new(channel);
    */

    // ===== Per-call token (combine with TLS) =====
    // Use an interceptor to attach the token so a fresh token rides on every call:
    let token = "my-secret-token";
    let token_val: MetadataValue<_> = format!("Bearer {}", token).parse()?;
    
    let auth_interceptor = move |mut req: tonic::Request<()>| {
        req.metadata_mut().insert("authorization", token_val.clone());
        Ok(req)
    };

    // let mut client_with_token = UserServiceClient::with_interceptor(channel, auth_interceptor);

    Ok(())
}
